plugins/package-managers/node/src/main/kotlin/pnpm/PnpmDependencyHandler.kt

```diff
@@ -0,0 +1,81 @@
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.plugins.packagemanagers.node.pnpm
+
+import java.io.File
+
+import org.ossreviewtoolkit.model.Identifier
+import org.ossreviewtoolkit.model.Issue
+import org.ossreviewtoolkit.model.Package
+import org.ossreviewtoolkit.model.PackageLinkage
+import org.ossreviewtoolkit.model.utils.DependencyHandler
+import org.ossreviewtoolkit.plugins.packagemanagers.node.PackageJson
+import org.ossreviewtoolkit.plugins.packagemanagers.node.parsePackage
+import org.ossreviewtoolkit.plugins.packagemanagers.node.parsePackageJson
+import org.ossreviewtoolkit.plugins.packagemanagers.node.pnpm.ModuleInfo.Dependency
+import org.ossreviewtoolkit.utils.common.realFile
+
+internal class PnpmDependencyHandler(private val pnpm: Pnpm) : DependencyHandler<Dependency> {
+    private val workspaceModuleDirs = mutableSetOf<File>()
+    private val packageJsonCache = mutableMapOf<File, PackageJson>()
+
+    private fun Dependency.isProject() = File(path).realFile().absoluteFile in workspaceModuleDirs
+
+    fun setWorkspaceModuleDirs(dirs: Collection<File>) {
+        workspaceModuleDirs.apply {
+            clear()
+            addAll(dirs)
+        }
+    }
+
+    override fun identifierFor(dependency: Dependency): Identifier {
+        val type = pnpm.managerName.takeIf { dependency.isProject() } ?: "NPM"
+        val namespace = dependency.from.substringBeforeLast("/", "")
+        val name = dependency.from.substringAfterLast("/")
+        val version = if (dependency.isProject()) {
+            readPackageJson(dependency.packageJsonFile).version.orEmpty()
+        } else {
+            dependency.version.takeUnless { it.startsWith("link:") || it.startsWith("file:") }.orEmpty()
+        }
+
+        return Identifier(type, namespace, name, version)
+    }
+
+    override fun dependenciesFor(dependency: Dependency): List<Dependency> =
+        (dependency.dependencies + dependency.optionalDependencies).values.toList()
+
+    override fun linkageFor(dependency: Dependency): PackageLinkage =
+        PackageLinkage.DYNAMIC.takeUnless { dependency.isProject() } ?: PackageLinkage.PROJECT_DYNAMIC
+
+    override fun createPackage(dependency: Dependency, issues: MutableCollection<Issue>): Package? =
+        dependency.takeUnless { it.isProject() }?.let {
+            parsePackage(
+                workingDir = it.workingDir,
+                packageJsonFile = it.packageJsonFile,
+                getRemotePackageDetails = pnpm::getRemotePackageDetails
+            )
+        }
+
+    private fun readPackageJson(packageJsonFile: File): PackageJson =
+        packageJsonCache.getOrPut(packageJsonFile.realFile()) { parsePackageJson(packageJsonFile) }
+}
+
+private val Dependency.workingDir: File get() = File(path)
+private val Dependency.packageJsonFile: File get() = File(path, "package.json")
```

Comment:
How about instead adding this to the class body of `Dependency`:

        @Transient
        val workingDir = File(path)

        @Transient
        val packageJsonFile = workingDir.resolve("package.json")

Reply:
I would like to separate upstream data model from any ORT properties for clarify. So, I prefer to keep these extensions.